pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,155 @@
+            "Iterated dict '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-dict",
+            "Emitted when items are added or removed to a dict being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+        "E4703": (
+            "Iterated set '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-set",
+            "Emitted when items are added or removed to a set being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            message_raised = self._modified_iterating_check(body_node, iter_obj)
+            if not message_raised:  # needs further check when message not raised
+                for child in body_node.get_children():
+                    self._modified_iterating_check(child, iter_obj)
+
+    def _modified_iterating_check(
+        self, node: nodes.NodeNG, iter_obj: nodes.NodeNG
+    ) -> bool:
+        msg_id = None
+        if self._modified_iterating_list_cond(node, iter_obj):
+            msg_id = "modified-iterating-list"
+        elif self._modified_iterating_dict_cond(node, iter_obj):
+            msg_id = "modified-iterating-dict"
+        elif self._modified_iterating_set_cond(node, iter_obj):
+            msg_id = "modified-iterating-set"
+        if msg_id is not None:
+            self.add_message(
+                msg_id,
+                node=node,
+                args=(iter_obj.name,),
+                confidence=interfaces.INFERENCE,
+            )
+            return True
+        return False
+
+    @staticmethod
+    def _is_node_expr_that_calls_attribute_name(node: nodes.NodeNG) -> bool:
+        return (
+            isinstance(node, nodes.Expr)
+            and isinstance(node.value, nodes.Call)
+            and isinstance(node.value.func, nodes.Attribute)
+            and isinstance(node.value.func.expr, nodes.Name)
+        )
+
+    @classmethod
+    def _common_cond_list_set(
+        cls,
+        node: nodes.Expr,
+        list_obj: nodes.NodeNG,
+        infer_val: Optional[nodes.NodeNG],
+    ) -> bool:
+        return (infer_val == utils.safe_infer(list_obj)) and (
+            node.value.func.expr.name == list_obj.name
+        )
+
+    @staticmethod
+    def _is_node_assigns_subscript_name(node: nodes.NodeNG) -> bool:
+        return isinstance(node, nodes.Assign) and (
+            isinstance(node.targets[0], nodes.Subscript)
+            and (isinstance(node.targets[0].value, nodes.Name))
+        )
+
+    @classmethod
+    def _modified_iterating_list_cond(
+        cls, node: nodes.NodeNG, list_obj: nodes.NodeNG
```

Comment:
```suggestion
        cls, node: nodes.NodeNG, iter_obj: nodes.NodeNG
```

We don't actually know if this is a list yet.